File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/audience_profile/compose_broadcast/ComposeBroadcastViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.audience_profile.compose_broadcast

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.audience.BroadcastHistoryMessageDto
import app.pantopus.android.data.api.models.audience.MembershipStatsCountsDto
import app.pantopus.android.data.api.models.audience.PersonaSummaryDto
import app.pantopus.android.data.api.models.audience.PublishUpdateBody
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.audience.AudienceProfileRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.time.Duration
import java.time.Instant
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import kotlin.math.roundToInt

/** Nav-arg key for the persona id read off the back-stack handle. */
const val COMPOSE_BROADCAST_PERSONA_ID_KEY = "personaId"

/**
 * A.7 (A22.2) — Backs the full-screen Compose Broadcast surface. The editor
 * fields (body / audience / media / schedule) plus the send [ComposePhase]
 * live in a single [ComposeBroadcastUiState]; the prompt's five-case contract
 * is derived via [ComposeBroadcastUiState.composeState].
 *
 * Wiring: [load] resolves the owner's persona + broadcast channel from
 * `GET /api/personas/me`, the per-tier reach from
 * `GET /api/personas/:id/membership-stats`, and the recent broadcasts from
 * `GET /api/broadcast/channels/:id/messages`. It also swaps [performSend] in
 * for the real publish (`POST .../messages`). The default [performSend] stays
 * a no-op success so unit tests drive the state machine without a network
 * (they never call [load]); [ComposeBroadcastSampleData] remains the
 * preview/snapshot seam.
 */
@HiltViewModel
class ComposeBroadcastViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
        private val repository: AudienceProfileRepository,
    ) : ViewModel() {
        private val personaId: String =
            savedStateHandle.get<String>(COMPOSE_BROADCAST_PERSONA_ID_KEY)
                ?.takeIf { it.isNotBlank() }
                ?: ComposeBroadcastSampleData.persona.id

        private var lastSavedDraft = ComposeBroadcastDraft()
        private var channelId: String? = null

        private val _state =
            MutableStateFlow(
                ComposeBroadcastUiState(
                    persona =
                        BroadcastPersona(
                            id = personaId,
                            handle = "",
                            displayName = "",
                            kind = ComposeBroadcastSampleData.persona.kind,
                            avatarInitial = "",
                        ),
                    recentBroadcasts = emptyList(),
                    draft = ComposeBroadcastDraft(),
                    scheduledAtMillis = null,
                    scheduledLabel = null,
                    phase = ComposePhase.Idle,
                    isDirty = false,
                    audienceReach = ComposeBroadcastSampleData.audienceReach,
                ),
            )
        val state: StateFlow<ComposeBroadcastUiState> = _state.asStateFlow()

        private val _recentsLoading = MutableStateFlow(false)
        val recentsLoading: StateFlow<Boolean> = _recentsLoading.asStateFlow()

        private val _recentsError = MutableStateFlow<String?>(null)
        val recentsError: StateFlow<String?> = _recentsError.asStateFlow()

        /** Stubbed network call. Default no-op success; [load] swaps in the
         *  real publish; tests override directly. */
        var performSend: suspend (ComposeBroadcastDraft, Long?) -> Unit = { _, _ -> }

        /** Resolve persona + channel, per-tier reach, and recent broadcasts,
         *  and wire the real publish. Best-effort: a failure on any leg leaves
         *  the composer usable; only the recents surface shows an error. */
        fun load() {
            performSend = { draft, _ -> realPublish(draft) }
            _recentsLoading.value = true
            _recentsError.value = null
            viewModelScope.launch {
                var resolvedPersonaId = personaId
                when (val meResult = repository.me()) {
                    is NetworkResult.Success -> {
                        meResult.data.persona?.let { summary ->
                            resolvedPersonaId = summary.id
                            _state.update { it.copy(persona = personaFrom(summary, it.persona)) }
                        }
                        channelId = meResult.data.channel?.id
                    }
                    is NetworkResult.Failure -> Unit // composer still works
                }

                when (val statsResult = repository.membershipStats(resolvedPersonaId)) {
                    is NetworkResult.Success ->
                        _state.update { it.copy(audienceReach = reachFrom(statsResult.data.counts)) }
                    is NetworkResult.Failure -> Unit
                }

                val channel = channelId
                if (channel != null) {
                    when (val historyResult = repository.broadcastHistory(channel)) {
                        is NetworkResult.Success ->
                            _state.update {
                                it.copy(recentBroadcasts = historyResult.data.messages.mapNotNull(::recentBroadcast))
                            }
                        is NetworkResult.Failure -> _recentsError.value = "Couldn't load recent broadcasts."
                    }
                }
                _recentsLoading.value = false
            }
        }

        private suspend fun realPublish(draft: ComposeBroadcastDraft) {
            val channel = channelId ?: error("Your broadcast channel isn't ready yet. Try again in a moment.")
            val (visibility, rank) = wireFor(draft.audience)
            val body =
                PublishUpdateBody(
                    body = draft.body.trim(),
                    visibility = visibility,
                    targetTierRank = rank,
                )
            when (val result = repository.publishUpdate(channel, body)) {
                is NetworkResult.Success -> Unit
                is NetworkResult.Failure -> throw result.error
            }
        }

        fun updateBody(text: String) = mutateDraft { it.copy(body = text) }

        fun setAudience(audience: BroadcastAudience) = mutateDraft { it.copy(audience = audience) }

        fun attachMedia(media: ComposeMediaPreview) = mutateDraft { it.copy(media = media) }

        fun removeMedia() = mutateDraft { it.copy(media = null) }

        fun schedule(atMillis: Long) {
            update {
                it.copy(scheduledAtMillis = atMillis, scheduledLabel = formatSchedule(atMillis))
            }
        }

        fun sendNow() {
            update { it.copy(scheduledAtMillis = null, scheduledLabel = null) }
        }

        fun saveDraft() {
            lastSavedDraft = _state.value.draft
            _state.update { it.copy(isDirty = false) }
        }

        fun retry() {
            if (_state.value.phase is ComposePhase.Error) {
                _state.update { it.copy(phase = ComposePhase.Idle) }
            }
        }

        @Suppress("TooGenericExceptionCaught")
        fun send(onSent: () -> Unit = {}) {
            val current = _state.value
            if (!current.canSend) return
            val snapshot = current.draft
            val at = current.scheduledAtMillis
            _state.update { it.copy(phase = ComposePhase.Sending) }
            viewModelScope.launch {
                try {
                    performSend(snapshot, at)
                    lastSavedDraft = ComposeBroadcastDraft(audience = snapshot.audience)
                    _state.update {
                        it.copy(
                            draft = lastSavedDraft,
                            scheduledAtMillis = null,
                            scheduledLabel = null,
                            phase = ComposePhase.Idle,
                            isDirty = false,
                        )
                    }
                    onSent()
                } catch (error: Throwable) {
                    _state.update {
                        it.copy(phase = ComposePhase.Error(error.message ?: "Couldn't send broadcast. Try again."))
                    }
                }
            }
        }

        private fun mutateDraft(transform: (ComposeBroadcastDraft) -> ComposeBroadcastDraft) {
            update { it.copy(draft = transform(it.draft)) }
        }

        /** Apply a state change, clearing any prior send error and recomputing dirty. */
        private fun update(transform: (ComposeBroadcastUiState) -> ComposeBroadcastUiState) {
            _state.update { current ->
                val next = transform(current)
                val phase = if (next.phase is ComposePhase.Error) ComposePhase.Idle else next.phase
                next.copy(
                    phase = phase,
                    isDirty = !next.draft.isEmpty && next.draft != lastSavedDraft,
                )
            }
        }

        private fun formatSchedule(millis: Long): String = SimpleDateFormat("MMM d, h:mm a", Locale.US).format(Date(millis))
    }

// MARK: - Pure mappers (mirror iOS `ComposeBroadcastViewModel` statics)

/** Targeting chip → broadcast `visibility` + `target_tier_rank`. */
internal fun wireFor(audience: BroadcastAudience): Pair<String, Int?> =
    when (audience) {
        BroadcastAudience.AllBeacons -> "public" to null
        BroadcastAudience.FollowersOnly -> "followers" to null
        BroadcastAudience.BronzePlus -> "tier_or_above" to 2
        BroadcastAudience.SilverPlus -> "tier_or_above" to 3
        BroadcastAudience.GoldOnly -> "tier_or_above" to 4
    }

/** Broadcast `visibility` + rank → targeting chip (inverse of [wireFor]). */
internal fun audienceFor(
    visibility: String?,
    rank: Int?,
): BroadcastAudience =
    when (visibility) {
        "public" -> BroadcastAudience.AllBeacons
        "followers" -> BroadcastAudience.FollowersOnly
        "tier_or_above", "subscribers" ->
            when (rank ?: 2) {
                in Int.MIN_VALUE..2 -> BroadcastAudience.BronzePlus
                3 -> BroadcastAudience.SilverPlus
                else -> BroadcastAudience.GoldOnly
            }
        else -> BroadcastAudience.AllBeacons
    }

@Suppress("MagicNumber")
internal fun reachFrom(counts: MembershipStatsCountsDto): Map<BroadcastAudience, Int> =
    mapOf(
        BroadcastAudience.AllBeacons to (counts.followers ?: 0),
        BroadcastAudience.FollowersOnly to (counts.followers ?: 0),
        BroadcastAudience.BronzePlus to (counts.members ?: 0),
        BroadcastAudience.SilverPlus to (counts.insiders ?: 0),
        BroadcastAudience.GoldOnly to (counts.direct ?: 0),
    )

internal fun personaFrom(
    summary: PersonaSummaryDto,
    fallback: BroadcastPersona,
): BroadcastPersona {
    val name = summary.displayName ?: summary.handle ?: fallback.displayName
    val handle = summary.handle?.let { "@$it" } ?: fallback.handle
    return fallback.copy(
        id = summary.id,
        handle = handle,
        displayName = name,
        avatarInitial = name.firstOrNull()?.uppercase() ?: fallback.avatarInitial,
    )
}

@Suppress("MagicNumber", "ReturnCount")
internal fun recentBroadcast(dto: BroadcastHistoryMessageDto): RecentBroadcastContent? {
    if (dto.locked == true) return null
    val id = dto.id ?: return null
    val delivered = dto.deliveredCount ?: 0
    val read = dto.readCount ?: 0
    val readPct = if (delivered > 0) "${(read.toDouble() / delivered * 100).roundToInt()}%" else "—"
    return RecentBroadcastContent(
        id = id,
        timeLabel = relativeTime(dto.publishedAt ?: dto.createdAt),
        audience = audienceFor(dto.visibility, dto.targetTierRank),
        body = dto.body ?: "",
        reach = shortCount(delivered),
        read = shortCount(read),
        readPct = readPct,
        // The broadcast serializer carries no reaction / reply counts.
        reactions = "—",
        replies = "—",
        hasMedia = !dto.media.isNullOrEmpty(),
    )
}

@Suppress("MagicNumber")
internal fun shortCount(count: Int): String =
    when {
        count < 1000 -> "$count"
        count < 10000 -> String.format(Locale.US, "%.1fK", count / 1000.0)
        else -> "${(count / 1000.0).roundToInt()}K"
    }

@Suppress("MagicNumber")
internal fun relativeTime(iso: String?): String {
    val instant = iso?.let { runCatching { Instant.parse(it) }.getOrNull() } ?: return ""
    val minutes = Duration.between(instant, Instant.now()).toMinutes()
    return when {
        minutes < 1 -> "Just now"
        minutes < 60 -> "${minutes}m ago"
        minutes < 1440 -> "${minutes / 60}h ago"
        minutes < 2880 -> "Yesterday"
        minutes < 10080 -> "${minutes / 1440}d ago"
        else -> "${minutes / 10080}w ago"
    }
}

File: frontend/apps/ios/Pantopus/Features/AudienceProfile/ComposeBroadcast/ComposeBroadcastViewModel.swift
//
//  ComposeBroadcastViewModel.swift
//  Pantopus
//
//  A.7 (A22.2) — Backs the full-screen Compose Broadcast surface. The
//  editor fields (body / audience / media / schedule) are stored and
//  bound directly; `state` is derived from them plus the send `phase`
//  so it always matches the prompt's
//  `.empty / .composing / .scheduled / .sending / .error` contract.
//
//  No backend: `performSend` defaults to a no-op success and can be
//  injected (latency / failure) by the host or tests. Recent broadcasts
//  + persona + per-audience reach are seeded from
//  `ComposeBroadcastSampleData`.
//

import Foundation
import Observation

@Observable
@MainActor
public final class ComposeBroadcastViewModel {
    /// Send lifecycle, kept separate from the editable draft so the
    /// draft survives a `.sending` / `.error` round-trip.
    enum Phase: Equatable {
        case idle
        case sending
        case error(String)
    }

    public let personaId: String
    public let persona: BroadcastPersona
    public private(set) var recentBroadcasts: [RecentBroadcastContent]
    public let maxCharacterCount: Int

    public private(set) var draft: ComposeBroadcastDraft
    public private(set) var scheduledAt: Date?
    private var phase: Phase = .idle
    private var lastSavedDraft: ComposeBroadcastDraft

    private let audienceReach: [BroadcastAudience: Int]
    private let onSent: @MainActor () -> Void

    /// Stubbed network call. Default is an immediate no-op success; the
    /// host injects latency and the test suite injects a thrower /
    /// state-capture closure. `var` so it can be swapped after init.
    var performSend: @MainActor (ComposeBroadcastDraft, Date?) async throws -> Void

    public init(
        personaId: String,
        persona: BroadcastPersona,
        recentBroadcasts: [RecentBroadcastContent],
        audienceReach: [BroadcastAudience: Int] = [:],
        draft: ComposeBroadcastDraft = ComposeBroadcastDraft(),
        scheduledAt: Date? = nil,
        maxCharacterCount: Int = 1000,
        onSent: @escaping @MainActor () -> Void = {},
        performSend: @escaping @MainActor (ComposeBroadcastDraft, Date?) async throws -> Void = { _, _ in }
    ) {
        self.personaId = personaId
        self.persona = persona
        self.recentBroadcasts = recentBroadcasts
        self.audienceReach = audienceReach
        self.draft = draft
        self.scheduledAt = scheduledAt
        self.maxCharacterCount = maxCharacterCount
        self.onSent = onSent
        self.performSend = performSend
        lastSavedDraft = draft
    }

    // MARK: - Derived state

    public var state: ComposeBroadcastState {
        switch phase {
        case .sending:
            return .sending
        case let .error(message):
            return .error(message: message)
        case .idle:
            if let sendAt = scheduledAt {
                return .scheduled(draft, sendAt: sendAt)
            }
            return draft.isEmpty ? .empty : .composing(draft)
        }
    }

    public var characterCount: Int {
        draft.body.count
    }

    public var isOverLimit: Bool {
        characterCount > maxCharacterCount
    }

    public var hasRecentBroadcasts: Bool {
        !recentBroadcasts.isEmpty
    }

    public var isSending: Bool {
        phase == .sending
    }

    /// Send is allowed once there's content, we're under the character
    /// limit, and a send isn't already in flight.
    public var canSend: Bool {
        guard phase != .sending else { return false }
        guard !draft.isEmpty else { return false }
        return !isOverLimit
    }

    /// Drives the top-bar unsaved-draft chip — content exists and it
    /// diverges from the last saved snapshot.
    public var isDirty: Bool {
        !draft.isEmpty && draft != lastSavedDraft
    }

    /// CTA copy flips to first-run when the persona has never broadcast.
    public var primaryActionTitle: String {
        hasRecentBroadcasts ? "Send broadcast" : "Send your first broadcast"
    }

    public func reach(for audience: BroadcastAudience) -> Int? {
        audienceReach[audience]
    }

    // MARK: - Editing intents

    public func updateBody(_ text: String) {
        draft.body = text
        recoverFromError()
    }

    public func setAudience(_ audience: BroadcastAudience) {
        draft.audience = audience
        recoverFromError()
    }

    public func attachMedia(_ media: ComposeMediaPreview) {
        draft.media = media
        recoverFromError()
    }

    public func removeMedia() {
        draft.media = nil
        recoverFromError()
    }

    public func schedule(at date: Date) {
        scheduledAt = date
        recoverFromError()
    }

    public func sendNow() {
        scheduledAt = nil
        recoverFromError()
    }

    /// Persist the current draft. No backend — we just snapshot it so the
    /// unsaved-draft chip clears.
    public func saveDraft() {
        lastSavedDraft = draft
    }

    /// Leave the error phase and return to the live composer.
    public func retry() {
        if case .error = phase { phase = .idle }
    }

    public func send() async {
        guard canSend else { return }
        let snapshot = draft
        let sendAt = scheduledAt
        phase = .sending
        do {
            try await performSend(snapshot, sendAt)
            // Reset the composer but keep the chosen audience as the
            // default for the next broadcast.
            draft = ComposeBroadcastDraft(audience: snapshot.audience)
            scheduledAt = nil
            lastSavedDraft = draft
            phase = .idle
            onSent()
        } catch {
            phase = .error(Self.message(for: error))
        }
    }

    // MARK: - Helpers

    /// Any edit clears a prior send error so the composer is live again.
    private func recoverFromError() {
        if case .error = phase { phase = .idle }
    }

    private static func message(for error: any Error) -> String {
        if let localized = error as? any LocalizedError, let description = localized.errorDescription {
            return description
        }
        return "Couldn't send broadcast. Try again."
    }
}
